Avoid port conflicts

dcdenu4 opened the pull request:
This PR attempts to solve the problem of launching the Flask App on a port that is already in use. It does this by:

- using a Python `socket` to check if a port is available. 
- If a port is not available, it tries the next port number for `n` amount of ports. 
- The client side which spawned the Flask App gets the used port number via listening to the spawned processes `stdout`.

This does NOT solve the problem of launching multiple Workbench UIs, which Dave and I discussed should be a separate issue.

dcdenu4 commented on src/main_helpers.js:
```diff
@@ -33,6 +33,7 @@ export function findInvestBinaries(isDevMode) {
       // If no dotenv vars are set, default to where this project's
       // build process places the binaries.
       serverExe = `${process.env.SERVER || 'build/invest/server'}${ext}`;
+      //serverExe = `${process.env.SERVER || 'build/invest/server'}`;
```
I left this commented code here, because when I was running in `dev` mode using `.env` and `server.py` the auto adding of `ext` was causing issues. Should the resolving line be:
```javascript
serverExe = `${process.env.SERVER}` || `build/invest/server${ext}`;
```

dcdenu4 replied:
Think I have it updated here: https://github.com/natcap/invest-workbench/pull/73/commits/b592fc7592d3d61a45d152369c75b2e25d92b10c

davemfish reviewed (changes requested):
Hey @dcdenu4 - thanks for working on this, it turned out to be a trickier problem than I realized back when the issue was created.

I have a few comments and suggestions here. They're all fairly minor - I think the general solution here is great. A few minor comments that I think might have been caught/warned about by `eslint`. Could you try that if you haven't already?

When the `server.py` changes get ported over to invest, we'll probably want to add to `test_cli.py` to cover the new command-line args you added here. 

Oh, and could you look into the failing `npm run test-flask-app`? It might just need a minor change since it also calls `createPythonFlaskProcess`

Thanks!

dcdenu4 commented on src/main.js:
```diff
@@ -36,7 +34,8 @@ const createWindow = async () => {
+  // Wait for a response from the server confirming the host information
+  await createPythonFlaskProcess(binaries.server, isDevMode);
   // Wait for a response from the server before loading the app
   await getFlaskIsReady();
```
Updated here: https://github.com/natcap/invest-workbench/pull/73/commits/addc71e847189ed79ebc81194592748915cad15c

dcdenu4 commented on src/main_helpers.js:
```diff
@@ -60,42 +61,64 @@ export function findInvestBinaries(isDevMode) {
+
+      logger.debug(`Started python process as PID ${pythonServerProcess.pid}`);
+      logger.debug(serverExe);
+      pythonServerProcess.stdout.on('data', (data) => {
```
Updated here with some other linting: https://github.com/natcap/invest-workbench/pull/73/commits/b358eef61b4524024505355b64f407c7abef1730

dcdenu4 commented on src/server.py:
```diff
@@ -209,4 +212,38 @@ def save_to_python():
+        port = port + 1
+
+    # Write to stdout so that parent process can receive the used port number
+    print(f'PORT {port}')
```
Yes! That makes much more sense then a print statement. Here: https://github.com/natcap/invest-workbench/pull/73/commits/4c398b9c396999ad1a1b085e7c1bb7f8c5b68362

dcdenu4 commented on src/server.py:
```diff
@@ -209,4 +212,38 @@ def save_to_python():
+        "--flex", default=True, help="The port number of the Flask Server.")
+    args = vars(parser.parse_args())
+    port = int(args.get('port', 5000))
+    flex = bool(args.get('flex', True))
```
Yep, definitely left over redundancy after iterating. Should be good here: https://github.com/natcap/invest-workbench/pull/73/commits/4c398b9c396999ad1a1b085e7c1bb7f8c5b68362

dcdenu4 commented on src/main_helpers.js:
```diff
@@ -60,42 +61,64 @@ export function findInvestBinaries(isDevMode) {
-      // also putting it's location on the PATH:
-      pythonServerProcess = spawn(path.basename(serverExe), {
-        env: { PATH: path.dirname(serverExe) },
+export async function createPythonFlaskProcess(serverExe, isDevMode) {
```
Oh yea, I do like that way better. Thanks! https://github.com/natcap/invest-workbench/pull/73/commits/2035fa1af648f9188637af70671395cc471bc55e

dcdenu4 commented:
Hey @davemfish Thanks for the feedback and helping me learn JS stuff. Tests are now passing. I added a `retries` optional variable to `createPythonFlaskProcess`, to allow GHA to have more time. I didn't follow all the `npx eslint` recommendations as some of them weren't very clear and some weren't related to code I changed, so I figured you had thought about, haha.

Let me know what you think!

davemfish reviewed (changes requested):
Hey @dcdenu4 , thanks for making those updates! I had some more comments on this round mainly because I'm also still learning how to do things properly in javascript and maybe didn't give great suggestions last time. Also, as we're discussing on the invest PR about the workbench-invest interface, I think it pays to really try to get this right up front. 

So there's one main comment about when & how to set the `process.env.PORT` variable. And another comment about whether we can send data from python to node via a stream other than stdout (so we don't have to check for the special PORT string _every_ time a stdout chunk comes in.

We can talk this stuff over and try to come up with a good solution together!

dcdenu4 commented:
@davemfish Just coming back to this PR after our long break. We had a pretty good long discussion about this and in the end I think we decided to shelf this? I believe we landed on maybe this port handling was over designing and that we should start by sticking with a hard-coded value and see if any issues come up?

davemfish commented:
> @davemfish Just coming back to this PR after our long break. We had a pretty good long discussion about this and in the end I think we decided to shelf this? I believe we landed on maybe this port handling was over designing and that we should start by sticking with a hard-coded value and see if any issues come up?

Yes, that's how I remember it! The stripped down changes that we decided on will go into PRs https://github.com/natcap/invest/pull/399 and #72 . So would love to get your review on those soon, but first I'm checking them over again and I'll post on those PRs when they're ready for review!

dcdenu4 commented:
Sounds good! I'll close this PR down.
